routes/api.js

```diff
@@ -238,58 +240,99 @@ var removeSession = function(sessionId) {
     return delete sessions[sessionId];
 };
 
-var proxyRoute = function(req, res, next) {
-    proxy.web(req, res);
+// Create a kernel. Does not use the proxy instance because it must parse the
+// and reserialize the request body with additional information in certain
+// configurations.
+router.post('/kernels', bodyParser.json({ type: 'text/plain' }), function(req, res) {
```

Comment:
@jtyberg and I tried numerous combinations of trying to keep everything as middleware, invoke `web.proxy` for all routes except `/api/kernels`, and invoke our special handling for `/api/kernels`. That approach grew really really complex because of the body parser middleware, modifications to the body, and restreaming of the modified body required for the proxy to them forward the modified to the backend. It was quite lengthy and we never did get it working. I think we were fighting a losing battle because all of the examples about intercepting the request body before proxying treat the body as immutable. In our case here, we want to modify the body to inject the auth info.

In the end, we went with having a plain old route handler for this one case where we want to mutate the request body.